### PR TITLE
Add ReadCausal Endpoint

### PR DESCRIPTION
**Pull request by arun-koshy**
This endpoint returns collections along a DAG walk with a well-defined starting point. More info can be found in [Issue#69](https://github.com/MystenLabs/narwhal/issues/69)

Few pending `TODOs` on this PR
- Currently integration test will fail when trying to read causal from a missing collection. This is because `PrimaryMessage::CertificatesBatchResponse` is not wired up properly to handle the fetched certificate. 
- Issue with resolving `test_utils` import. More info can be found in slack [question](https://mysten-labs.slack.com/archives/C02L5JL88JV/p1652946318171069)
- Currently only returning the first error returned by `block_synchronizer_handler.get_and_synchronize_block_headers()`, should we be returning each individual sync error?

### CODE REVIEW

File: primary/src/grpc_server/validator.rs [hunk 1]
```diff
@@ -14,35 +19,89 @@ use tonic::{Request, Response, Status};
...
+
+        let block_header_results = self
+            .block_synchronizer_handler
+            .get_and_synchronize_block_headers(ids.clone())
+            .await;
+
+        for result in block_header_results {
+            if result.is_err() {
+                // TODO: Should we capture & return each individual sync error?
+                return Err(Status::internal(format!(
```
Comment (akichidis) [inline primary/src/grpc_server/validator.rs:82]: I would say returning just one error for now is ok. 

You can also write the statement as:

```
if let Err(err) = result {
    return Err(Status::internal(format!("Error when trying to synchronize block headers: {:?}", err)));
}
```

so you don't have to deal with `unwrap_err` etc

File: primary/tests/integration_tests.rs [hunk 2]
```diff
@@ -384,6 +385,192 @@ async fn test_remove_collections() {
...
+    // Test read causal for missing collection from Primary 1. Expect block synchronizer
+    // to handle retrieving the missing collections from Primary 2 before completing the
+    // request for read causal.
+    let request = tonic::Request::new(ReadCausalRequest {
+        collection_id: Some(collection_ids[0].into()),
+    });
+
+    let status = client.read_causal(request).await.unwrap_err();
+
+    // TODO: Test should succeed when PrimaryMessage::CertificatesBatchResponse is handled properly.
```
[the remark this answers is not in the document]
Comment (akichidis) [reply primary/tests/integration_tests.rs:568]: I did run the test and enabled the logs to see the issue. So it seems that the problem is not with handling the `PrimaryMessage::CertificateBatchResponse` message, but with the replied certificate from the peer. The following log shows that the certificate is invalid:
```
2022-05-19T10:26:47.816796Z ERROR primary::block_synchronizer::responses: Found at least one invalid certificate from peer Ed25519PublicKey(PublicKey(CompressedEdwardsY: [99, 28, 21, 65, 243, 164, 191, 68, 212, 216, 151, 6, 21, 100, 170, 132, 149, 215, 102, 246, 25, 26, 63, 246, 21, 98, 0, 63, 24, 75, 140, 101]), EdwardsPoint{
	X: FieldElement51([1583941683691414, 1106778313907963, 451323164995954, 397979000143795, 435135634094729]),
	Y: FieldElement51([2151689512885347, 739777969817751, 452304450005392, 293548659117307, 1786451675312134]),
	Z: FieldElement51([1, 0, 0, 0, 0]),
	T: FieldElement51([1852706767764721, 1865025095120952, 2241810975302701, 1361639220993947, 492103237434098])
})). Will ignore all certificates
```

This happens when the validation of the certificate fails: https://github.com/MystenLabs/narwhal/blob/4bbbd6e58bb86edcff570bbe03ad209bf72bfa58/primary/src/block_synchronizer/responses.rs#L116
in this case the certificate is ignored and is not fetched back.

Should have a look on the way that the mock certificates are created and ensure that valid ones are produced.

File: primary/src/grpc_server/validator.rs [hunk 3]
```diff
@@ -14,35 +19,89 @@ use tonic::{Request, Response, Status};
...
+    Validator for NarwhalValidator<PublicKey, SynchronizerHandler>
+{
+    async fn read_causal(
+        &self,
+        request: Request<ReadCausalRequest>,
+    ) -> Result<Response<ReadCausalResponse>, Status> {
+        let collection_id = request
+            .into_inner()
+            .collection_id
+            .ok_or_else(|| Status::invalid_argument("Missing collection"))?;
```
Comment (akichidis) [inline primary/src/grpc_server/validator.rs:71]: nit: I would probably make the error message a bit more descriptive, something like:
```
No collection id has been provided
```

**Review by akichidis — COMMENTED** [review]
Great work @arun-koshy ! Please have a look on my comments around the integration test

File: primary/src/grpc_server/validator.rs [hunk 4]
```diff
@@ -14,35 +19,89 @@ use tonic::{Request, Response, Status};
...
+            }
+        }
+
+        if let Some(dag) = &self.dag {
+            let result = match dag.read_causal(ids[0]).await {
+                Ok(digests) => Ok(ReadCausalResponse {
+                    collection_ids: digests
+                        .iter()
+                        .map(|digest| digest.to_owned().into())
+                        .collect::<Vec<CertificateDigestProto>>(),
```
[the remark this answers is not in the document]
Comment (arun-koshy) [reply primary/src/grpc_server/validator.rs:95]: Nice one! Thanks

File: primary/tests/integration_tests.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (huitseeker) [reply primary/tests/integration_tests.rs:568]: Hey, you know what, I kinda want to make sure we keep this exact integration test in, but with an expected failure, showing that we do validate the certificates :D 
We should test we accept valid certs mocked in a better way, but the negative test is too useful to throw away.

[the remark this answers is not in the document]
Comment (akichidis) [reply primary/tests/integration_tests.rs:568]: Good idea on the negative test! 👍

**Review by huitseeker — COMMENTED** [review]
This looks good - I'll work on getting you a better (signed) mock certificate.

File: consensus/src/tests/consensus_tests.rs [hunk 5]
```diff
@@ -2,140 +2,18 @@
...
-    parents: BTreeSet<CertificateDigest>,
-) -> (CertificateDigest, Certificate<Ed25519PublicKey>) {
-    let certificate = Certificate {
-        header: Header {
-            author: origin,
-            round,
-            parents,
-            ..Header::default()
-        },
-        ..Certificate::default()
```
[the remark this answers is not in the document]
Comment (huitseeker) [reply consensus/src/tests/consensus_tests.rs:51]: This is fixed in #265

File: consensus/src/tests/consensus_tests.rs [hunk 6]
```diff
@@ -2,140 +2,18 @@
...
-    initial_parents: &BTreeSet<CertificateDigest>,
-    keys: &[Ed25519PublicKey],
-) -> (
-    VecDeque<Certificate<Ed25519PublicKey>>,
-    BTreeSet<CertificateDigest>,
-) {
-    make_certificates(start, stop, initial_parents, keys, 0.0)
-}
-
-pub fn make_certificates(
```
[the remark this answers is not in the document]
Comment (huitseeker) [reply consensus/src/tests/consensus_tests.rs:71]: Introduced another mock to get valid certificates in https://github.com/MystenLabs/narwhal/pull/265

File: primary/tests/integration_tests.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (arun-koshy) [reply primary/tests/integration_tests.rs]: Thanks for the help with this! Included both the negative & positive test.

**Pull request merged**